openai/client.py

```diff
@@ -0,0 +1,392 @@
+import time
+import typing
+
+import openai
+
+LATEST_AZURE_API_VERSION = "2023-05-15"
+
+
+class AzureTokenAuth:
+    """Authentication using an Azure AD token.
+    """
+
+    def __repr__(self):
+        return f"AzureTokenAuth({type(self._credential)})"
+
+    def __init__(self, *, credential=None):
+        """Create a new AzureTokenAuth instance. If no credential is passed, 
+        it will use ~azure.identity.DefaultAzureCredential
+        """
+        if not credential:
+            try:
+                import azure.identity
+            except ImportError:
+                raise Exception(
+                    "You have to install the azure-identity package in order to use AzureTokenAuth"
+                )
+            credential = azure.identity.DefaultAzureCredential()
+        self._credential = credential
+        self._cached_token = None
+
+    def get_token(self) -> str:
+        if (
+            self._cached_token is None
+            or (self._cached_token.expires_on - time.time()) < 300
+        ):
+            self._cached_token = self._credential.get_token(
+                "https://cognitiveservices.azure.com/.default"
+            )
+
+        return self._cached_token.token
+
+
+class ApiKeyAuth:
+    """Authentication using an API key"""
+
+    def __repr__(self):
+        return f"ApiKeyAuth(api_key=<redacted>)"
+    
+    def __init__(self, key: str = ""):
+        """Create a new ApiKeyAuth instance. If no key is passed, it will use ~openai.api_key"""
+        self.key = key or openai.api_key
+
+    def get_token(self) -> str:
+        """Get the API key"""
+        return self.key
+
+
+Backends = typing.Literal["azure", "openai", ""]
+
+
+class OpenAIClient:
+    def __init__(
```

Comment:
I think we need to support passing `api_key_path` and `organization` as well

Reply:
Hm maybe just `organization` actually. Doesn't seem like `api_key_path` is passed through

Reply:
Agreed, yes, organization should be squirreled away.

Reply:
Done.